### DISCUSSION TITLE
More Best Practice

### DISCUSSION THREAD

**Discussion by Bodyclock-Games** [stride3d/stride#1981] [Q&A]
Hi All, I'm continuing with my terrain tool conversion and am pivoting towards the entity / component / processor approach. Before I head further down this rabbit hole I just wanted to confirm I'm on the right track.

So I have a WorldManger component that holds all data for the world and manages the tiles. This creates a World service. I tried using the Settings system and added custom settings and created a service from there, which worked fine, but the settings interface was already crowded and the world settings were going to get quite large once the procedural data is added. It would be nice if there was a custom blank settings available to use.

So I then created a Tile component which holds specific local data (position etc) and a Tile processor. See below:

```
namespace Bodyclock.World.Terrain
{
    internal class TileProcessor : EntityProcessor<Tile, WorldManager>, IEntityComponentRenderProcessor
    {
        public VisibilityGroup VisibilityGroup { get; set; }
        private WorldManager World { get; set; }

        protected override void OnSystemAdd()
        {
            base.OnSystemAdd();
            World = Services.GetService<WorldManager>();
        }

        protected override WorldManager GenerateComponentData([NotNull] Entity entity, [NotNull] Tile component)
        {
            return World;
        }

        public override void Draw(RenderContext context)
        {
            base.Draw(context);
            foreach (var data in ComponentDatas)
            {
                if (data.Key.IsDirty)
                {
                    var mesh = Mesh.GenerateMeshData(data.Value, data.Key);
                    data.Key.IsDirty = false;
                }
            }
        }
    }
}
```

This incomplete but works fine and will produce the mesh data for the Tiles that need it.

So a couple of questions; Firstly this loads the World Service at OnSystemAdd. This works but I'm concerned that it was luck that the service was populated before I called it. Is there a way to set load priority in code?
And in general, is this a resonable approach. Any comments and improvements welcomed.
Edit: Oh, and is it a good idea to make the GenerateMeshData method async?
Cheers
Chris

**Comment by manio143** [comment]
Looking at your processor - you're not really using the tiles components in there - sounds to me like you'd be better off with a custom GameSystem instead.

Use a processor whenever you need to process a component.

On the note of having a custom settings asset - what you're describing sounds a lot like a custom asset - I'm slowly working on the plugin system for Stride which will hopefully make those possible.

**Reply by Bodyclock-Games** [reply]
Thanks for the reply. My bad, I will be using the Tile component in the final version, this was more to get the concept right. The data object in the collection is the component and that will be updated with the created mesh. This is mainly used in the LOD system. I should have made that clearer. I will have a look at GameSystems also.

Given that, would the ComponentDatas loop be the place for an async function to update the tiles?

And yes, the settings idea was what I thought might be a quick way to get a scriptable data object. I've looked at your Custom Asset project which looks good, a version of Unity's ScriptableObject which are extremly useful beyond their original scope, even if they are somewhat abused at times.
